scylla-proxy/src/proxy.rs

```diff
@@ -681,22 +683,27 @@ impl Doorkeeper {
         real_addr: SocketAddr,
         shard_awareness: ShardAwareness,
     ) -> Result<TcpStream, DoorkeeperError> {
-        let cluster_stream = if shard_awareness.is_aware() {
-            let shards = match self.shards_number {
-                None => {
-                    let temporary_stream = TcpStream::connect(real_addr)
-                        .await
-                        .map_err(|err| DoorkeeperError::NodeConnectionAttempt(real_addr, err))?;
-                    let shards = self
-                        .obtain_shards_number(temporary_stream, real_addr)
-                        .await?;
-                    self.shards_number = Some(shards);
-
-                    shards
-                }
-                Some(shards) => shards,
-            };
+        let shards = if shard_awareness.is_aware() {
+            if let shards @ Some(_) = self.shards_number {
+                shards
+            } else {
+                let temporary_stream = TcpStream::connect(real_addr)
```

Comment:
Do I understand correctly that, in case of Cassandra nodes, `make_cluster_stream` will keep issuing probe connections because `self.shards_number` will keep being `None`? The driver can be configured to establish multiple connections to the same node, even with a Cassandra cluster (it doesn't do it by default though, save for the control connection). It's not a big issue, but one that is probably easy to avoid.

Besides, why doesn't the Doorkeeper try to establish the number of shards at the moment when it's created?

Reply:
Your right, and so I've moved querying the number of shards to the beginning of Doorkeepers operation.